lib/WP_Auth0_LoginManager.php

```diff
@@ -163,136 +170,124 @@ public function init_auth0() {
 
-	    $data->id_token = null;
-	    $response = WP_Auth0_Api_Client::get_user(
-		    $this->a0_options->get( 'domain' ),
-		    WP_Auth0_Api_Client::get_client_token(),
-		    $decoded_token->sub
-	    );
+      // Unsuccessful for another reason
+      if ( $exchange_resp instanceof WP_Error ) {
+        WP_Auth0_ErrorManager::insert_auth0_error( __METHOD__ . ' L:' . __LINE__, $exchange_resp );
+      }
 
-      if ( $response instanceof WP_Error ) {
-        WP_Auth0_ErrorManager::insert_auth0_error( __METHOD__ . ' => WP_Auth0_Api_Client::get_user()', $response );
+      throw new WP_Auth0_LoginFlowValidationException( __( 'Unknown error', 'wp-auth0' ), $exchange_resp_code );
+    }
 
-        error_log( $response->get_error_message() );
+    $data = json_decode( $exchange_resp_body );
 
-        throw new WP_Auth0_LoginFlowValidationException( );
-      }
+    if ( empty( $data->access_token ) ) {
 
-      $userinfo = json_decode( $response['body'] );
-      if ( $this->login_user( $userinfo, $data->id_token, $data->access_token ) ) {
-        if ( ! empty( $state_decoded->interim ) ) {
-          include WPA0_PLUGIN_DIR . 'templates/login-interim.php';
-          exit();
-        } else {
-          if ( ! empty( $state_decoded->redirect_to ) && wp_login_url() !== $state_decoded->redirect_to ) {
-            $redirectURL = $state_decoded->redirect_to;
-          } else {
-            $redirectURL = $this->a0_options->get( 'default_login_redirection' );
-          }
+      // Look for clues as to what went wrong
+      throw new WP_Auth0_LoginFlowValidationException(
+        ! empty( $data->error_description ) ? $data->error_description : __( 'Unknown error', 'wp-auth0' ),
+        ( ! empty( $data->error ) ? $data->error : $exchange_resp_code )
+      );
+    }
 
-          wp_safe_redirect( $redirectURL );
-        }
-      }
-    } elseif ( is_array( $response['response'] ) &&  401 === (int) $response['response']['code'] ) {
+    $decoded_token = JWT::decode(
+      $data->id_token,
+      $this->a0_options->get_client_secret_as_key(),
+      array( $this->a0_options->get_client_signing_algorithm() )
+    );
 
-      $error = new WP_Error( '401', 'auth/token response code: 401 Unauthorized' );
+    $userinfo_resp = WP_Auth0_Api_Client::get_user(
+      $this->a0_options->get( 'domain' ),
+      WP_Auth0_Api_Client::get_client_token(),
+      $decoded_token->sub
+    );
 
-      WP_Auth0_ErrorManager::insert_auth0_error( __METHOD__ . ' => $this->login_user() = 401', $error );
+    $userinfo_resp_code = (int) wp_remote_retrieve_response_code( $userinfo_resp );
+    $userinfo_resp_body = wp_remote_retrieve_body( $userinfo_resp );
 
-      $msg = __( 'Error: the Client Secret configured on the Auth0 plugin is wrong. Make sure to copy the right one from the Auth0 dashboard.', 'wp-auth0' );
+    if ( $userinfo_resp instanceof WP_Error || 200 !== $userinfo_resp_code || empty( $userinfo_resp_body ) ) {
```

Comment:
To be honest I think the "error handler" should catch the last 2 conditions already and create/raise a WP_ERROR. You should only be checking for the error existence, if not you are "extending" the error handler logic to every call.

Reply:
Also a good call. If `$userinfo_resp` is a `WP_Error` then this will also be true (resolves the code to 0).